### PR TITLE
Perf: Improve perf of `getCurrentPageShapesSorted`

### PR DESCRIPTION
**Pull request by MitjaBezensek**
This significantly improves performance. Here's a comparison with 2k shapes. Top is the new logic, bottom the old one.

image

One place where this does make a significant difference is when you have a lot of shapes on the page and you start [creating a new arrow](https://github.com/orgs/tldraw/projects/40?pane=issue&itemId=59296136):

Before:
image
image

After:
image
image

### Change Type

- [ ] `sdk` — Changes the tldraw SDK
- [ ] `dotcom` — Changes the tldraw.com web app
- [ ] `docs` — Changes to the documentation, examples, or templates.
- [ ] `vs code` — Changes to the vscode plugin
- [x] `internal` — Does not affect user-facing stuff

- [ ] `bugfix` — Bug fix
- [ ] `feature` — New feature
- [x] `improvement` — Improving existing features
- [ ] `chore` — Updating dependencies, other boring stuff
- [ ] `galaxy brain` — Architectural changes
- [ ] `tests` — Changes to any test code
- [ ] `tools` — Changes to infrastructure, CI, internal scripts, debugging tools, etc.
- [ ] `dunno` — I don't know

### CODE REVIEW

File: packages/editor/src/lib/editor/Editor.ts [hunk 1]
```diff
@@ -4573,31 +4573,38 @@ export class Editor extends EventEmitter<TLEventMap> {
...
-		// todo: consider making a derivation or something, or merging with rendering shapes
 		const shapes = new Set(this.getCurrentPageShapes().sort(sortByIndex))
-
-		const results: TLShape[] = []
-
-		function pushShapeWithDescendants(shape: TLShape): void {
-			results.push(shape)
-			shapes.delete(shape)
-
-			shapes.forEach((otherShape) => {
```
Comment (MitjaBezensek) [inline packages/editor/src/lib/editor/Editor.ts:4586]: This was the problematic part. For each shape we added we went through all the other shapes to find it's children. Now we have a map of parent child relationships which gets generated on the first pass through the shapes.

**Comment by steveruizok** [comment]
That's an error from the 1 password extension btw

File: packages/editor/src/lib/editor/Editor.ts (hunk 1 above)
Comment (steveruizok) [reply packages/editor/src/lib/editor/Editor.ts:4586]: damn, good one!

**Comment by steveruizok** [comment]
did a few little hot path tricks that we've used elsewhere. It's setting around .35ms

**Pull request merged**